### ISSUE TITLE
Mismatched lines

### ISSUE THREAD

**Issue by GitsHubber** [pnedev/comparePlus#352]
Lines from one document seem to be matched/compared to lines from another that they shouldn't compared to. That is to say that there are lines which are a far better/closer match. Also sometimes they are matched/compared to lines which are not much of a match, and infact have no equivalent line in the other document.  

Perhaps better explained with a screenshot example. 

https://images2.imgbox.com/30/7e/3tNRGzNT_o.png

**Comment by thereddestdog** [comment]
I think this is also related.. just a simple text compare:

![image](https://user-images.githubusercontent.com/24947579/229542825-35d79dd7-d8dd-4230-a337-350327e25334.png)

**Comment by pnedev** [comment]
@GitsHubber ,

Could you please show a screenshot of the same compare but without hiding the matched lines (that is without `Show Only Diffs` option)? Could you also restore the default coloring just for the screenshot?
I suspect that in between the diff sections you also have matched sections and the changed lines are looked for and marked only between the corresponding diff sections. So even if there is a better match for your changed line elsewhere in the file it depends on the diff context (the positioning of the line in the file corresponding to the other one).
This screenshot should make things more clear:

![changed_lines](https://user-images.githubusercontent.com/6064913/229742984-5fa99994-cb0c-4aac-8902-b152823ee311.png)

Here line 1 is better matched to line 3 and vice versa but because of line 2 those are separated and cannot be linked. The lines order between files has also changed.

@thereddestdog ,

You example looks good to me, what do you consider wrong there?

BR

**Comment by thereddestdog** [comment]
@pnedev I think the extra inserted lines are not correct. I think it should look like this:

![image](https://user-images.githubusercontent.com/24947579/230158205-f8bcb6c7-ba2a-4d3f-9e0b-ea3a9b67c494.png)

**Comment by pnedev** [comment]
@thereddestdog ,

Thank you for the feedback. You are right - old Compare plugin and the new ComparePlus plugin use different approaches when finding changed lines. There is no one universal solution and criteria on what is the right outcome (it depends on the text itself) - it is more intuitive for us people and we also have certain expectations.

For example look at these screenshots that illustrate the different approaches in another case:

![compareplus110](https://user-images.githubusercontent.com/6064913/230374342-5184d995-cbd1-4500-a08e-77a21ea4f65a.png)
![compare202](https://user-images.githubusercontent.com/6064913/230374346-97453db5-9070-4232-bbff-fdfc43d6f78f.png)

For your example ComparePlus gives a "right" result but it kind-of stresses out that lines 10-11 are moved before lines 8-9 (so lines 8-9 and 10-11 have changed places) and are also changed in the beginning - R1, R2 are the diffs and the rest of the lines contents are the context (sort-of-speak).

Compare plugin also gives a "right" result claiming that no lines have changed places (lines order is the same) only the end of the lines are different (cu and op210) and the rest are the context (sort-of-speak).

Which result is better cannot be determined with certainty - it depends on the context and on which difference matters most.

Here is an example with your files data but slightly modified:

![compareplus110](https://user-images.githubusercontent.com/6064913/230379650-f6fac023-bfd9-4763-bf24-61c517803671.png)
![compare202](https://user-images.githubusercontent.com/6064913/230379661-3729adff-b443-495d-a506-1523abeb3607.png)

BR

**Comment by Yaron10** [comment]
Hello Pavel,

We've discussed this issue before.
ComparePlus is absolutely brilliant, and as you wrote: "Which result is better cannot be determined with certainty - it depends on the context and on which difference matters most.".

Allow me to seize the opportunity and bring another example from my recent real-life Compare. 

Could you please compare [these files](https://github.com/pnedev/comparePlus/files/11183420/columnEditor.zip)?
Go to `case IDC_COL_INITNUM_EDIT`.

_Result:_

![תמונה](https://user-images.githubusercontent.com/10229320/230723573-05b453fe-eaa1-412a-8f2d-4a6f29e62094.png)

I've tried several on-line Compare sites, _and most indeed display that section as CP does_.
But I've come across https://onlinetextcompare.com/ whose display seems clearer _in this case_.

![תמונה](https://user-images.githubusercontent.com/10229320/230724134-087a8a7d-1425-4d5d-a967-6a349110ab98.png)

I do understand the issue is quite complex, and ComparePlus' approach is certainly much better in many cases.
So, this is just another example of this issue. :)

Thank you very much for your ongoing work. Much appreciated. 👍 

BR

**Comment by pnedev** [comment]
Hello my friend Yaron,

I appreciate your feedback as always.
Detecting the most "appealing" to us changed lines pairs is an ambiguous task indeed. And I am still not totally happy with its implementation as well.

Great example! 👍 
I have tried to make ComparePlus more customizable and to be able to accept "guidance" from us as to what we would like to stress in a comparison.
ComparePlus is by default punctual and pays attention to the empty lines as part of the compare context so in your example files to achieve the same result as https://onlinetextcompare.com/ you should set the `Changed threshould percentage` in the plugin Settings to 20% (or less) and also to use the `Ignore Empty Lines` option. That way you as a user decide to exclude the empty lines as something important and remove them from the compare context so the plugin tries to "look for differences" that are more important to you in that particular compare.

Happy holidays!
BR,
Pavel

**Comment by Yaron10** [comment]
Hello Pavel,

I hope you had a good day.

> And I am still not totally happy with its implementation as well.

**_I_** am very happy! :) 
And obviously - like in every field in life, there's always room for some improvements.

I've mentioned it several times before:
just analyzing the various possible scenarios and the ultimate desired result is a major task. 
_Implementing_ it is a mammoth project. I truly appreciate and _admire_  your work.

It's been a few years now, but I have never fully and thoroughly analyzed the best Compare results.
Please remember that when reading the following comment. :)

I set `Min line resemblance to mark as changed` to `20%` a long time ago.
Checking `Ignore Empty Lines`, the `case IDC_COL_INITNUM_EDIT` section is indeed displayed very clearly.
The "ultimate desired result" in this case. 👍 

Could you please look at two more sections?

**1.** `case IDC_COL_DEC_RADIO`.

![תמונה](https://user-images.githubusercontent.com/10229320/230792210-bb718623-56d4-480f-8d9e-5184556e969a.png)

![תמונה](https://user-images.githubusercontent.com/10229320/230792315-4b69b781-30f9-406c-9fbd-eff79fcd2554.png)

**2.** `case CBN_SELCHANGE`.

![תמונה](https://user-images.githubusercontent.com/10229320/230792579-339ae682-b0a6-4263-8cf1-c33c56ede328.png)

![תמונה](https://user-images.githubusercontent.com/10229320/230792616-a4ed3cb4-2ffe-4d32-a8d5-c982f318cfed.png)

Thank you very much and best wishes.

**Comment by pnedev** [comment]
Hello again Yaron,

Thank you.
You are right about the new examples in your last post - I noticed them as well and they seem "sub-optimal" but I need to analyze them thoroughly and if they could be improved. I currently don't have the time to do so. They are now expected because of the
`{
break;`
constructs that give recognizable match compare context.

I will write back if I come with some improvement on this.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Whenever you find time to look into it and possibly revise some elements in the current approach, it would be highly appreciated.

Analyzing only the few examples in this thread -
The ultimate goal is letting the user grasp the diffs _as quickly, easily and clearly as possible_.
Inserting new (alignment) lines, and thus creating "steps" - is a **major** slowing/confusing factor. It seems that NOT creating "steps" whenever possible would be much better.

***
![תמונה](https://user-images.githubusercontent.com/10229320/230900725-85e33f46-81d3-4c66-89d4-eef119837057.png)

vs.

![תמונה](https://user-images.githubusercontent.com/10229320/230900432-cae05b95-8cfa-47c3-801e-629bf583b9ef.png)

I do understand the logic and advantage in the first display, but the second one is simpler more easily "grasped".

***
![תמונה](https://user-images.githubusercontent.com/10229320/230905035-40dd8dd5-2270-4a84-8adc-80110e38f17b.png)

The logic of this smart detection and display is obvious.
But I'm still wondering if it's worth the confusing "steps".
Perhaps even here marking lines `2` & `4` as "not-equal" and lines `3` as "added/removed" should be better. 

***
![תמונה](https://user-images.githubusercontent.com/10229320/230907739-44cf6288-c285-4228-bf4a-0a9ad2512993.png)

Again, the logic of this smart detection and display is obvious.
But is it worth extra-lines/steps?

***

Bottom line:
It's quite complex.

Thanks again for your time and hard work. 👍

**Comment by pnedev** [comment]
Hi Yaron,

Thanks again for the feedback and your thoughts on it, I appreciate it.
I'll think about a way to improve this and make the changed detection decision even "smarter" but I won't have much free time in the near future so whenever I get to it.
I'll currently focus mainly on fixing major issues with ComparePlus although there are some really good and useful feature requests and I also have some ideas for additional functionalities.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Whenever you have free time it'll be appreciated.

> I'll currently focus mainly on fixing major issues with ComparePlus

What "major issues"? :)
Do you mean this thread?
_I don't mean to rush you whatsoever_ - but the more I think about it, the more important it seems.

Allow me a few more points regarding the current issue.
I know that suggesting is much easier than implementing.
Thank you for your patience. 

***

![תמונה](https://user-images.githubusercontent.com/10229320/231208891-92b23b0d-1051-4de7-ae07-958a352044d4.png)

I've mentioned the extra-lines/steps as a potential slowing/confusing factor.
That is: the user has to analyze **6** lines displayed in "steps" instead of "straight" **4**.

Another factor is the lines numbers.
1. We don't play with shuffled cards.
The fact that two lines, *although different*, are positioned in line (e.g.) `8` in both files is important, and should (IMO) be taken into account by the "smart detection".
 
2. Keeping lines `8` aligned should also make the user grasp the diffs more quickly.

***

The different lines background colors is also a major slowing factor.
That is: the "smart detection" should prefer to flow with the same block-background-color when possible.

![תמונה](https://user-images.githubusercontent.com/10229320/231220118-8f22f7fc-3f8d-44f0-ad2d-21df191ffa9f.png)

***

> although there are some really good and useful feature requests and I also have some ideas for additional functionalities.

👍 
I'll be looking forward to them.
Thank you so much. I do appreciate it.

**Comment by pnedev** [comment]
Hello Yaron,

> You wrote you didn't have free time these days, ....... Please let me know if you'd rather get back to it some time in the future.

Its OK, I don't mind discussing how to further improve ComparePlus, it's a pleasure :)
I just won't have much time soon to implement new, complex features or changes.

> Still, please allow me to ask: are you also referring to the case IDC_COL_INITNUM_EDIT, case IDC_COL_DEC_RADIO & case CBN_SELCHANGE sections?

IDC_COL_INITNUM_EDIT is OK as it is now because if we ignore the empty lines the compare context changes and the comparison is as expected.
As to IDC_COL_DEC_RADIO and CBN_SELCHANGE - those are compared OK now (by that I mean algorithmically it is expected and accurate) but from user perspective it is not optimal for sure and there is currently no way to "guide" the compare process to the desired compare context. Changing that will be complex but I'll think about making things better (when I find the time and energy).

> I wanted to emphasize that different background colors can also be visually distracting...

Yes, I agree. I am not sure though that keeping the background green/red only will be better. Maybe it is worth trying and experimenting with it.

> Even if the label="Accounts" lines should be marked as "changed", it might be better (in this case) to mark them as "added/removed" so that the we have only two background colors in the block.

Unfortunately I cannot do much about that. The only possibility would be to provide a way for the user to disable changed lines matching process. Or to use green/red background for changed lines as mentioned above.

Thank you for the feedback and sharing your thoughts about change detection implementation. :+1: 

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> Its OK, I don't mind discussing how to further improve ComparePlus, it's a pleasure :)

Great. It's a pleasure for me as well.

> but from user perspective it is not optimal for sure and there is currently no way to "guide" the compare process to the desired compare context. Changing that will be complex but I'll think about making things better (when I find the time and energy).

👍 
Time, energy and mood.
I do appreciate it.

> I am not sure though that keeping the background green/red only will be better.

I may not have been clear about that, but I did not suggest it categorically.
On the contrary, the "tricolor" in general is useful and helpful.

> Unfortunately I cannot do much about that...

That point was not as important.

Thank you. 👍 
I wish you the best. - Energy and mood. :)

**Comment by pnedev** [comment]
Hello Yaron,

I have "optimized" compare diffs in the same dev binaries pointed in https://github.com/pnedev/comparePlus/issues/353#issuecomment-1513139107. For convenience I'll point to the same dev binaries here: **[win32](https://ci.appveyor.com/api/buildjobs/nuj73295vwu97pp7/artifacts/ComparePlus.dll)**, **[win64](https://ci.appveyor.com/api/buildjobs/vvbqel581pt7ap0y/artifacts/ComparePlus.dll)**
Now if you ignore the empty lines you would get the same result with your example files as when using https://onlinetextcompare.com/. That was a very helpful feedback, thank you for sharing it :+1: 
I haven't come across such compare example before.
I probably still need to do some optimizations (in speed mainly, the need for those might not be user-noticeable at the moment) but the dev binaries are perfectly usable as they are now.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Thank you for this *major improvement*. Both the `columnEditor` pair and the `Chrome` pair look "more correct", and the diffs are much easier to grasp. 👍 

I appreciate your time and great work.

If you currently don't have time (or energy) for further discussing it, would you mind keeping this issue open?

Before the recent changes, the following section was "correctly" aligned *even if `Ignore Empty Lines` was unchecked*. 

![תמונה](https://user-images.githubusercontent.com/10229320/232859333-669243d2-a268-48c0-bf3f-70dd7f5ec09e.png)

With the recent commit:

![תמונה](https://user-images.githubusercontent.com/10229320/232860732-1e83430d-df45-47ec-a3b9-7937d33848fd.png)

Don't you think the first results are better regardless of `Ignore Empty Lines`?
It's like pulling a short blanket. :)

***

> I probably still need to do some optimizations (in speed mainly, the need for those might not be user-noticeable at the moment) but the dev binaries are perfectly usable as they are now.

Thank you for that too and best of luck.

**Comment by pnedev** [comment]
Hello Yaron,

Thank you.
I intend to keep this thread open for some time.

> Don't you think the first results are better regardless of Ignore Empty Lines?

They are better for sure, you are right. But I need to analyze the compare algorithm thoroughly to check if and how something can be further improved.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

I know that analyzing the compare algorithm should require a lot of and work.
Whenever you find the time and energy, It would be highly appreciated. 👍 

BR

**Comment by pnedev** [comment]
Hello all,

I have finally released ComparePlus v3.0.0 (https://github.com/pnedev/comparePlus/releases/tag/cp_3.0.0) that includes major compare algorithms rework in both accuracy and speed.

BR
